### PR TITLE
database: Optimize IterateRepoGitserverStatus

### PR DESCRIPTION
**Pull request by efritz**
Reduces query times of `IterateRepoGitserverStatus` from 11s (rare) / 9s (often) -> 10s (rare) / 3.5s (often). This is in the top 10 queries contributing to database load according to query insights.

*Insight*: Breaking the `LEFT JOIN` into an `INNER JOIN` and a `NOT EXISTS` query that when combined forms the same result set has good performance benefits. It allows us to use indexes that are otherwise not available due to spurious filter conditions. The `UNION ALL` is especially nice since we don't care about order and each of the subqueries are non-intersecting (thus we neither need a sort nor a gather to reduce duplicates, both which are expensive in memory, and usually by extension latency).

### CODE REVIEW

File: internal/database/gitserver_repos.go [hunk 1]
```diff
@@ -83,25 +83,14 @@ func (s *GitserverRepoStore) IterateRepoGitserverStatus(ctx context.Context, opt
 		return errors.New("nil repoFn")
 	}
 
-	q := `
```
Comment (efritz) [inline internal/database/gitserver_repos.go:86]: Without condition:

```
127.0.0.1 sg@sg ❯ explain analyze SELECT repo.id,
       repo.name,
       gr.clone_status,
       gr.shard_id,
       gr.last_external_service,
       gr.last_error,
       gr.last_fetched,
       gr.updated_at
FROM repo
    LEFT JOIN gitserver_repos gr ON gr.repo_id = repo.id
    WHERE repo.deleted_at IS NULL;
                                                                             QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Merge Left Join  (cost=7.23..991800.08 rows=3697965 width=270) (actual time=0.031..10971.253 rows=3814103 loops=1)
   Merge Cond: (repo.id = gr.repo_id)
   ->  Index Only Scan using a_good_repo_deleted_at_idx on repo  (cost=0.43..664826.75 rows=3695562 width=52) (actual time=0.023..3174.893 rows=3814103 loops=1)
         Heap Fetches: 3430734
   ->  Index Scan using gitserver_repos_pkey on gitserver_repos gr  (cost=0.43..270458.13 rows=6863335 width=222) (actual time=0.005..5517.859 rows=6857469 loops=1)
 Planning Time: 1.843 ms
 Execution Time: 11187.460 ms
(7 rows)
```

With condition:

```
127.0.0.1 sg@sg ❯ explain analyze SELECT repo.id,
       repo.name,
       gr.clone_status,
       gr.shard_id,
       gr.last_external_service,
       gr.last_error,
       gr.last_fetched,
       gr.updated_at
FROM repo
    LEFT JOIN gitserver_repos gr ON gr.repo_id = repo.id
    WHERE repo.deleted_at IS NULL AND (gr.shard_id = '' OR gr IS NULL);
                                                                                   QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=1007.23..949994.38 rows=18490 width=270) (actual time=8971.759..8974.024 rows=0 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Merge Left Join  (cost=7.23..947145.38 rows=7704 width=270) (actual time=8967.353..8967.358 rows=0 loops=3)
         Merge Cond: (repo.id = gr.repo_id)
         Filter: ((gr.shard_id = ''::text) OR (gr.* IS NULL))
         Rows Removed by Filter: 1271368
         ->  Parallel Index Only Scan using a_good_repo_deleted_at_idx on repo  (cost=0.43..643269.30 rows=1539818 width=52) (actual time=0.039..1152.735 rows=1271368 loops=3)
               Heap Fetches: 3430733
         ->  Index Scan using gitserver_repos_pkey on gitserver_repos gr  (cost=0.43..270458.13 rows=6863335 width=468) (actual time=0.047..6572.937 rows=6857410 loops=3)
 Planning Time: 0.379 ms
 Execution Time: 8974.071 ms
(12 rows)

Time: 8991.883 ms (00:08.992)
```

File: internal/database/gitserver_repos.go [hunk 2]
```diff
@@ -147,6 +136,36 @@ FROM repo
 	return nil
 }
 
+const iterateRepoGitserverStatusQuery = `
```
Comment (efritz) [inline internal/database/gitserver_repos.go:139]: Without condition:

```
TODO - optimize, this is slower
```

With condition:

```
                                                                                         QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 HashAggregate  (cost=1020101.97..1020286.88 rows=18491 width=156) (actual time=4113.214..4197.507 rows=0 loops=1)
   Group Key: repo.id, repo.name, (NULL::text), (NULL::text), (NULL::bigint), (NULL::text), (NULL::timestamp with time zone), (NULL::timestamp with time zone)
   ->  Gather  (cost=1000.43..1019732.15 rows=18491 width=156) (actual time=4113.050..4197.342 rows=0 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Parallel Append  (cost=0.43..1016883.05 rows=18491 width=156) (actual time=4077.822..4077.828 rows=0 loops=3)
               ->  Parallel Hash Anti Join  (cost=182272.20..835164.93 rows=1 width=172) (actual time=3273.751..3273.755 rows=0 loops=3)
                     Hash Cond: (repo.id = gr.repo_id)
                     ->  Parallel Index Only Scan using a_good_repo_deleted_at_idx on repo  (cost=0.43..643269.30 rows=1539818 width=52) (actual time=0.059..1304.958 rows=1271369 loops=3)
                           Heap Fetches: 3430739
                     ->  Parallel Hash  (cost=146525.23..146525.23 rows=2859723 width=4) (actual time=1157.847..1157.848 rows=2285824 loops=3)
                           Buckets: 8388608  Batches: 1  Memory Usage: 333856kB
                           ->  Parallel Seq Scan on gitserver_repos gr  (cost=0.00..146525.23 rows=2859723 width=4) (actual time=0.012..412.742 rows=2285824 loops=3)
               ->  Nested Loop  (cost=0.43..181440.75 rows=7704 width=270) (actual time=1206.101..1206.102 rows=0 loops=2)
                     ->  Parallel Seq Scan on gitserver_repos gr_1  (cost=0.00..153674.54 rows=14299 width=222) (actual time=1206.095..1206.095 rows=0 loops=2)
                           Filter: ((shard_id = ''::text) OR (gr_1.* IS NULL))
                           Rows Removed by Filter: 3428736
                     ->  Index Only Scan using a_good_repo_deleted_at_idx on repo repo_1  (cost=0.43..1.94 rows=1 width=52) (never executed)
                           Index Cond: (id = gr_1.repo_id)
                           Heap Fetches: 0
 Planning Time: 0.618 ms
 Execution Time: 4198.088 ms
(22 rows)
```

File: internal/database/gitserver_repos.go [hunk 3]
```diff
@@ -83,22 +83,11 @@ func (s *GitserverRepoStore) IterateRepoGitserverStatus(ctx context.Context, opt
 		return errors.New("nil repoFn")
 	}
 
-	q := `
```
Comment (efritz) [inline internal/database/gitserver_repos.go:86]: Without condition:

```
                                                                             QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Merge Left Join  (cost=7.23..991800.08 rows=3697965 width=270) (actual time=0.025..10924.190 rows=3814133 loops=1)
   Merge Cond: (repo.id = gr.repo_id)
   ->  Index Only Scan using a_good_repo_deleted_at_idx on repo  (cost=0.43..664826.75 rows=3695562 width=52) (actual time=0.018..3168.190 rows=3814133 loops=1)
         Heap Fetches: 3430772
   ->  Index Scan using gitserver_repos_pkey on gitserver_repos gr  (cost=0.43..270458.13 rows=6863335 width=222) (actual time=0.004..5505.639 rows=6857488 loops=1)
 Planning Time: 1.809 ms
 Execution Time: 11136.705 ms
(7 rows)

Time: 11183.182 ms (00:11.183)
```

With condition:

```
                                                                                   QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=1007.23..949994.38 rows=18490 width=270) (actual time=9928.617..9930.094 rows=0 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Merge Left Join  (cost=7.23..947145.38 rows=7704 width=270) (actual time=9923.573..9923.574 rows=0 loops=3)
         Merge Cond: (repo.id = gr.repo_id)
         Filter: ((gr.shard_id = ''::text) OR (gr.* IS NULL))
         Rows Removed by Filter: 1271378
         ->  Parallel Index Only Scan using a_good_repo_deleted_at_idx on repo  (cost=0.43..643269.30 rows=1539818 width=52) (actual time=0.038..1345.142 rows=1271378 loops=3)
               Heap Fetches: 3430772
         ->  Index Scan using gitserver_repos_pkey on gitserver_repos gr  (cost=0.43..270458.13 rows=6863335 width=468) (actual time=0.052..7323.467 rows=6857417 loops=3)
 Planning Time: 0.397 ms
 Execution Time: 9930.132 ms
(12 rows)

Time: 9948.038 ms (00:09.948)
```

File: internal/database/gitserver_repos.go [hunk 4]
```diff
@@ -147,6 +136,52 @@ FROM repo
 	return nil
 }
 
+const iterateRepoGitserverQuery = `
+-- source: internal/database/gitserver_repos.go:GitserverRepoStore.IterateRepoGitserverStatus
+SELECT
```
Comment (efritz) [inline internal/database/gitserver_repos.go:141]:
```
                                                                             QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Merge Left Join  (cost=7.23..991800.08 rows=3697965 width=270) (actual time=0.016..10719.117 rows=3814136 loops=1)
   Merge Cond: (repo.id = gr.repo_id)
   ->  Index Only Scan using a_good_repo_deleted_at_idx on repo  (cost=0.43..664826.75 rows=3695562 width=52) (actual time=0.009..3082.242 rows=3814136 loops=1)
         Heap Fetches: 3430772
   ->  Index Scan using gitserver_repos_pkey on gitserver_repos gr  (cost=0.43..270458.13 rows=6863335 width=222) (actual time=0.004..5409.720 rows=6857491 loops=1)
 Planning Time: 0.248 ms
 Execution Time: 10930.321 ms
(7 rows)

Time: 10948.700 ms (00:10.949)
```

File: internal/database/gitserver_repos.go [hunk 5]
```diff
@@ -147,6 +136,52 @@ FROM repo
...
+	gr.last_fetched,
+	gr.updated_at
+FROM repo
+LEFT JOIN gitserver_repos gr ON gr.repo_id = repo.id
+WHERE repo.deleted_at IS NULL
+`
+
+const iterateRepoGitserverStatusWithoutShardQuery = `
+-- source: internal/database/gitserver_repos.go:GitserverRepoStore.IterateRepoGitserverStatus
+(
```
Comment (efritz) [inline internal/database/gitserver_repos.go:157]:
```
                                                                                      QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=1000.43..989842.35 rows=2 width=221) (actual time=3802.044..3877.577 rows=0 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Parallel Append  (cost=0.43..988842.15 rows=2 width=221) (actual time=3770.237..3770.243 rows=0 loops=3)
         ->  Parallel Hash Anti Join  (cost=182272.20..835164.93 rows=1 width=172) (actual time=3455.459..3455.463 rows=0 loops=3)
               Hash Cond: (repo.id = gr.repo_id)
               ->  Parallel Index Only Scan using a_good_repo_deleted_at_idx on repo  (cost=0.43..643269.30 rows=1539818 width=52) (actual time=0.066..1317.781 rows=1271381 loops=3)
                     Heap Fetches: 3430774
               ->  Parallel Hash  (cost=146525.23..146525.23 rows=2859723 width=4) (actual time=1234.674..1234.675 rows=2285832 loops=3)
                     Buckets: 8388608  Batches: 1  Memory Usage: 333824kB
                     ->  Parallel Seq Scan on gitserver_repos gr  (cost=0.00..146525.23 rows=2859723 width=4) (actual time=0.014..402.630 rows=2285832 loops=3)
         ->  Nested Loop  (cost=0.43..153677.18 rows=1 width=270) (actual time=472.162..472.164 rows=0 loops=2)
               ->  Parallel Seq Scan on gitserver_repos gr_1  (cost=0.00..153674.54 rows=1 width=222) (actual time=472.161..472.161 rows=0 loops=2)
                     Filter: (shard_id = ''::text)
                     Rows Removed by Filter: 3428748
               ->  Index Only Scan using a_good_repo_deleted_at_idx on repo repo_1  (cost=0.43..2.65 rows=1 width=52) (never executed)
                     Index Cond: (id = gr_1.repo_id)
                     Heap Fetches: 0
 Planning Time: 0.533 ms
 Execution Time: 3877.654 ms
(20 rows)

Time: 3917.859 ms (00:03.918)
```

**Review by eseliger — APPROVED** [review]

**Pull request merged**